sonic_db_config/db_config.go

```diff
@@ -1,135 +1,237 @@
-    if !sonic_db_init {
-        DbInit()
-    }
-    db_list := GetDbList()
-    separator, ok := db_list[db_name].(map[string]interface{})["separator"]
-    if !ok {
-        panic(fmt.Errorf("'separator' is not a valid field in database_config.json file!"))
-    }
-    return separator.(string)
-}
-
-func GetDbId(db_name string)(int) {
-    if !sonic_db_init {
-        DbInit()
-    }
-    db_list := GetDbList()
-    id, ok := db_list[db_name].(map[string]interface{})["id"]
-    if !ok {
-        panic(fmt.Errorf("'id' is not a valid field in database_config.json file!"))
-    }
-    return int(id.(float64))
-}
-
-func GetDbSock(db_name string)(string) {
-    if !sonic_db_init {
-        DbInit()
-    }
-    inst := GetDbInst(db_name)
-    unix_socket_path, ok := inst["unix_socket_path"]
-    if !ok {
-        panic(fmt.Errorf("'unix_socket_path' is not a valid field in database_config.json file!"))
-    }
-    return unix_socket_path.(string)
-}
-
-func GetDbHostName(db_name string)(string) {
-    if !sonic_db_init {
-        DbInit()
-    }
-    inst := GetDbInst(db_name)
-    hostname, ok := inst["hostname"]
-    if !ok {
-        panic(fmt.Errorf("'hostname' is not a valid field in database_config.json file!"))
-    }
-    return hostname.(string)
-}
-
-func GetDbPort(db_name string)(int) {
-    if !sonic_db_init {
-        DbInit()
-    }
-    inst := GetDbInst(db_name)
-    port, ok := inst["port"]
-    if !ok {
-        panic(fmt.Errorf("'port' is not a valid field in database_config.json file!"))
-    }
-    return int(port.(float64))
-}
-
-func GetDbTcpAddr(db_name string)(string) {
-    if !sonic_db_init {
-        DbInit()
-    }
-    hostname := GetDbHostName(db_name)
-    port := GetDbPort(db_name)
-    return hostname + ":" + strconv.Itoa(port)
+func GetDbDefaultNamespace() string {
+	return SONIC_DEFAULT_NAMESPACE
+}
+func CheckDbMultiNamespace() bool {
+	if !sonic_db_init {
+		DbInit()
+	}
+	return sonic_db_multi_namespace
+}
+func GetDbNonDefaultNamespaces() ([]string, int) {
+	if !sonic_db_init {
+		DbInit()
+	}
+	ns_list := make([]string, 0, len(sonic_db_config))
+	for ns := range sonic_db_config {
+		if ns == SONIC_DEFAULT_NAMESPACE {
+			continue
+		}
+		ns_list = append(ns_list, ns)
+	}
+	return ns_list, len(ns_list)
+}
+func GetDbAllNamespaces() []string {
+	if !sonic_db_init {
+		DbInit()
+	}
+	ns_list := make([]string, len(sonic_db_config))
+	i := 0
+	for ns := range sonic_db_config {
+		ns_list[i] = ns
+		i++
+	}
+	return ns_list
+}
+
+func GetDbNamespaceFromTarget(target string) (string, bool) {
+	if target == GetDbDefaultNamespace() {
+		return target, true
+	}
+	ns_list, _ := GetDbNonDefaultNamespaces()
+	for _, ns := range ns_list {
+		if target == ns {
+			return target, true
+		}
+	}
+	return "", false
+}
+func GetDbList(ns string) map[string]interface{} {
+	if !sonic_db_init {
+		DbInit()
+	}
+	db_list, ok := sonic_db_config[ns]["DATABASES"].(map[string]interface{})
```

Comment:
Is it a possible case that target ns is not present in the map?
```
db_list, ok := sonic_db_config[ns]
ok == false
```

Reply:
that is not possible. we read name space from database_config.json file so if that does not contain ns we will hit some other error before.